Add UniformCylindricalFlatEndcap.

markscheel opened the pull request:
## Proposed changes

Another map necessary for a binary compact object domain decomposition using cylinders.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

markscheel committed:
Add UniformCylindricalFlatEndcap.

markscheel commented:
Fixed a few codecov things.

The clang-tidy errors are apparently all toms748.

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.hpp:
```diff
@@ -0,0 +1,526 @@
+ * 1 that has \f$z \geq z_{\mathrm{P}1}\f$, and on this portion of the
+ * sphere the angular coordinate \f$\theta_1 = \acos((z-C_1^2)/R_1)\f$
+ * is uniform in \f$\bar{\rho} = \sqrt{\bar{x}^2+\bar{y}^2}\f$ and the angular
+ * coordinate \f$\phi_1 = \atan((y-C_1^1)/(x-C_1^0))\f$ is the same as
```
Same comment as the `UniformCylindricalSide` PR about using $x, y, z$ for the labels: $C_1^1 \rightarrow C_1^y$.

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.hpp:
```diff
@@ -0,0 +1,526 @@
+ * according to
+ *
+ * \f{align}
+ * x^0 &= C_1^0+\lambda(C_2^0-C_1^0) +
```
Same here: $x^0, x^1, x^2 \rightarrow x, y, z$

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.hpp:
```diff
@@ -0,0 +1,526 @@
+ *      \left(x^0-C_1^0-\lambda(C_2^0-C_1^0)\right)(C_2^0-C_1^0)+
+ *      \left(x^1-C_1^1-\lambda(C_2^1-C_1^1)\right)(C_2^1-C_1^1)
+ *      \right]\nonumber \\
+ *     &
+ *    -2 \left((1-\lambda)R_1\sin\theta_1 +
```
I think this should be a `+2` instead of a `-2`.

markscheel replied:
I still get -2.  This -2 comes from differentiating the last term in eq 10, which has a minus sign in front.

knelli2 replied:
You are correct. Differentiated it again and got -2 this time. Must have made a mistake the first time. Oops!

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.hpp:
```diff
@@ -0,0 +1,526 @@
+ * It is expected that `inverse()` will often be passed points
+ * \f$(x,y,z)\f$ that are out of the range of the map; in this case
+ * `inverse()` returns a default-constructed
+ * `std::optional<std::array<double, 3>>`. To avoid the difficulty and
```
`std::nullopt` here and below

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.hpp:
```diff
@@ -0,0 +1,526 @@
+ * Finally, consider the circle \f$S_1\f$ defining the intersection of sphere 1
+ * and the plane \f$z = z_{\mathrm{P}1}\f$; this circle has radius
+ * \f$r_1 = R_1 \sin\theta_{1\mathrm{max}}\f$.
+ * Now consider the cone that passes through both \f$S_1\f$ and
+ * the circle \f$S_3\f$ bounding the upper disk.
```
Doesn't $S_2$ bound the uppder disk with radius $R_2$ and center $C_2$? Or is $S_3$ a different circle?

markscheel replied:
Yeah it is S_2.  S_3 was a typo.

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.hpp:
```diff
@@ -0,0 +1,526 @@
+ * \f$\alpha\f$ is shown in the figure above. If
+ * \f$\alpha < \theta_{1 \mathrm{max}}\f$, then the line segment \f$L'\f$
+ * intersects the mapped portion of sphere 1 twice, so the map is
+ * multiply valued. Therefore we demand that the map parameters are such that
```
`multi-valued`

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.hpp:
```diff
@@ -0,0 +1,526 @@
+ * everywhere.  We cannot numerically check every point in this
+ * one-parameter family of points, but we demand that this condition
+ * is satisfied for a reasonably large number of points in this
+ * family.  This check is not very expensive since it is done only
```
Resonably large being 100? 1k? 10k? `std::numeic_limits<size_t>::max()`?

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.hpp:
```diff
@@ -0,0 +1,526 @@
+ * one-parameter family of points, but we demand that this condition
+ * is satisfied for a reasonably large number of points in this
+ * family.  This check is not very expensive since it is done only
+ * once, in the constructor.
+ *
+ */
```
Mention somewhere that this map can only be used in 3D.

markscheel replied:
Added another '3D' in the \brief

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.hpp:
```diff
@@ -0,0 +1,526 @@
+ *
+ * We also demand that the z plane in the above figure lies
+ * above the center of the sphere and is not too close to the center
+ * or edge of the sphere; specificially, we demand that
```
`specifically`

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,692 @@
+  // roundoff.  So ignore that point, and start at the next point.
+  // Keep track of the sign, and look for sign changes as we progress
+  // from point to point.  zero is a sentinel value.
+  double sign = 0.0;
```
Maybe `std::optional<double>` would be better here? Just to avoid confusion with over-assigning meaning to `0.0`.

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,692 @@
+        ASSERT(abs(cos_theta_max) < 1.0,
+               "Plane one must intersect sphere_one, and at more than one "
+               "point. You probably specified a bad value of z_plane_one, "
+               "radius_one, or the z component of center_one.");
```
List the values that were given?

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,692 @@
+  // as long as the unit test is changed to test them.
+  ASSERT(center_two[2] >= center_one[2] + 1.05 * radius_one,
+         "center_two[2] must be >= center_one[2] + 1.02 * radius_one, not "
+             << center_two[2] << " " << center_one[2] + 1.02 * radius_one);
```
`1.02` -> `1.05`

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,692 @@
+             << center_two[2] << " " << center_one[2] + 5.0 * radius_one);
+  ASSERT(radius_two >= 0.1 * radius_one * sin(theta_max_one_),
+         "radius_two is too small: " << radius_two << " "
+                                     << 0.1 * radius_one * sin(theta_max_one_));
```
I don't know if this condition is listen in the docs?

knelli2 replied:
I still think this condition isn't covered in the docs.

However, upon closer inspection, I think this condition is similar to the $\alpha > 1.1 \theta_{1 \mathrm{max}}$ condition because it's a restriction on how small `radius_two` can be in relation to the size of the sphere. I would either add a comment here in the cpp file about it so we don't get confused later on why it's here, or just explicitly state the condition in the docs.

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.hpp:
```diff
@@ -0,0 +1,526 @@
+ * (1-\bar{z})R_1 S(\bar{\rho},\theta_{1 \mathrm{max}}) +
+ * (1+\bar{z})R_2
+ *     \right), \label{eq:x1alt} \\
+ * x^2 &= \frac{1}{2}\left((1-\bar{z})C_1^2 + (1+\bar{z})C_2^2\right) +
+ * (1-\bar{z})R_1 \cos\theta_1, \label{eq:x2alt} \\
```
Factor of 0.5 on the last term?

markscheel replied:
Yup, good catch.

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,692 @@
+  // check z >= z_plane_one_
+  if (target_coords[2] < z_plane_one_ and
+      not equal_within_roundoff(target_coords[2], z_plane_one_)) {
+    return {};
```
`return std::nullopt;`

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,692 @@
+            << rhobar_min << " " << function_at_rhobar_min << " " << rhobar_max
+            << " " << function_at_rhobar_max << " " << new_rhobar_min << " "
+            << function_at_new_rhobar_min << " " << deriv_function_at_rhobar_min
+            << " " << new_rhobar_min - rhobar_min << "\n");
```
Label the values in the error messages, here and below.

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,692 @@
+  get<1, 1>(jac) = get<2, 0>(jac) + 0.5 * radius_two_ * (1.0 + zbar);
+  get<2, 1>(jac) = theta_max_one_ * get<2, 0>(jac);
+  get<1, 2>(jac) =
+      0.5 * radius_two_ - 0.5 * radius_one_ *
```
Factor out 0.5 so you only multiply it once? I don't actually know if it's more efficient or not.

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,692 @@
+
+bool operator==(const UniformCylindricalFlatEndcap& lhs,
+                const UniformCylindricalFlatEndcap& rhs) {
+  // don't need to compare theta_max_one_ or theta_max_two_
```
We don't have a `theta_max_two_` in this class

knelli2 commented on tests/Unit/Domain/CoordinateMaps/Test_UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,167 @@
+  // returns an invalid std::optional when called for a point that is
+  // outside the range of the map.
+
+  // Point with z less than z_plane_one.
+  CHECK_FALSE(map.inverse({{0.0,0.0,z_plane_one-1.0}}));
```
`.has_value()` to be explicit

knelli2 reviewed (changes requested):
Full review.

markscheel commented:
Made changes.  In a separate commit, made another change to print more info in all of the ASSERTS in the constructor, which helps for debugging.

knelli2 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,712 @@
+
+bool operator==(const UniformCylindricalFlatEndcap& lhs,
+                const UniformCylindricalFlatEndcap& rhs) {
+  // don't need to compare theta_max_one_
+  // because is is uniquely determined from the other variables.
```
`is is` -> `it is`

knelli2 reviewed (changes requested):
Looks good! Just one last change. Please squash everything together

knelli2 commented:
Also @geoffrey4444 can you please take a look at this?

markscheel commented:
Made last small change and squashed.

knelli2 reviewed (commented):
Ok everything else looks good. One more thing I overlooked when reviewing the fixups.

geoffrey4444 commented on tests/Unit/Domain/CoordinateMaps/Test_UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,171 @@
+  };
+  const double invertible_at_zero = is_invertible(0.0);
+  const double invertible_at_max = is_invertible(rho_max_so_far);
+  ASSERT(invertible_at_zero == 1.0, "Map should always be invertible at rho=0");
```
Why is this test an ASSERT instead of a CHECK? If you're running the test, don't you always want to check this?

geoffrey4444 commented on tests/Unit/Domain/CoordinateMaps/Test_UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,171 @@
+                       {{center_one[0] + rho, center_one[1], z_center_two}},
+                       radius_one, radius_two, theta_max_one)
+               ? 1.0
+               : -1.0;
```
Why are you using doubles instead of bools? I don't see how you could have roundoff error or anything with what's being done here, but it just seems kind of strange

geoffrey4444 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,712 @@
+      sign = func > 0.0 ? 1.0 : -1.0;
+    } else if (func * sign.value() <= 0.0) {
+      // There is another root, so function is not invertible.
+      return false;
```
Do you understand why codecov is mad about this? Would it be hard to modify the unit test to handle this case? Same comment for all codecov failures

geoffrey4444 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,712 @@
+    return buffer.str();
+  };
+
+  // Assumptions made in the map.  Some of these can be relaxed,
```
If any can't be relaxed, worth noting that in the comment here?

geoffrey4444 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,712 @@
+                                radius_one_)) {
+    // sqrt above because we don't want to change the scale of
+    // equal_within_roundoff too much.
+    return {};
```
What's the difference between `return {}` and return `std::nullopt`, if any? If none, why not be consistent?

markscheel replied:
Oops. Those should have been std::nullopt.  Changed.

(there is no difference.  `return {}` and `return std::nullopt` both construct an invalid std::optional).

geoffrey4444 commented on src/Domain/CoordinateMaps/UniformCylindricalFlatEndcap.cpp:
```diff
@@ -0,0 +1,712 @@
+  // function that is zero when rhobar has the correct value, and the
+  // derivative of that function with respect to rhobar.
+  const auto function_and_deriv_to_zero = [this, &target_coords](
+                                              const double rhobar) {
```
Why doesn't this function just call `this_function_is_zero_for_correct_rhobar`?

markscheel replied:
I just removed this function because it is no longer needed as-is.  Only a function that computes the deriv is needed, which I added instead of this function.

geoffrey4444 reviewed (changes requested):
Some minor comments for you to address and then I'm happy to approve. My apologies for taking so long to take a loot @markscheel !

markscheel commented:
I think I have addressed everything asked for.

geoffrey4444 commented:
LGTM! I don't need to see this again before merging, but I'm still curious about the one place where you use 1.0 and -1.0 instead of a bool

nilsdeppe reviewed (commented):
I have no additional comments

markscheel commented:
Squashed.

I added a comment to the code about 1.0 and -1.0 vs bool.   That function is used in a root finder that expects doubles, so it cannot be a bool.  (The root finder finds the value of rho where the function value changes from 1.0 to -1.0, i.e. where the function fails to be invertible).

knelli2 commented:
@markscheel Needs a rebase now that #4247 is merged.

markscheel commented:
Rebased.
